src/algebra/group/mod.rs

```diff
@@ -0,0 +1,71 @@
+#![doc = include_str!("./README.md")]
+pub mod prime;
+
+use std::ops::{Add, AddAssign, Mul, MulAssign, Neg, Sub, SubAssign};
+
+use super::Finite;
+
+#[const_trait]
+/// Group trait defined by a binary operation, identity element and inverse.
+pub trait Group:
+  std::fmt::Debug
+  + Sized
+  + Add<Output = Self>
+  + AddAssign
+  + Sub<Output = Self>
+  + SubAssign
+  + Neg
+  + Mul<Self::Scalar, Output = Self>
+  + MulAssign<Self::Scalar>
+  + Clone
+  + Copy
+  + Default
+  + Eq {
+  /// Scalar defined in the group
+  type Scalar;
+  /// Identity element of group
+  const IDENTITY: Self;
+
+  /// operation defined for the group, can be `+` for additive group and `·` for multiplicative
+  /// group
+  fn op(&self, rhs: &Self) -> Self;
+
+  /// Inverse of group element: a·i = [`FiniteGroup::IDENTITY`]
+  fn inverse(&self) -> Option<Self>;
+
+  /// Multiplication with the scalar element of the group, i.e. repeatedly applying group
+  /// [`FiniteGroup::operation`] `scalar` number of times.
+  fn scalar_mul(&self, scalar: Self::Scalar) -> Self;
+}
+
+/// Group trait with finite number of elements
+pub trait FiniteGroup: Finite + Group {
+  /// order of group element, i.e. order of finite cyclic subgroup that the element belongs to
+  fn order(&self) -> usize {
+    let mut order = 1;
+    let mut elem = *self;
+    for _ in 0..Self::ORDER {
+      // check if elem is the identity
+      if elem == Self::IDENTITY {
+        return order;
+      }
+      // apply operation and increment order
+      elem = Self::op(&elem, self);
+      order += 1;
+    }
+    order
+  }
+}
+
+/// Defines a group with commutative operation: `a·b=b·a`
+pub trait AbelianGroup: Group {
+  /// Returns whether the group is an abelian group
```

Comment:
A little misleading. This returns true if a pair of elements commutes. 

Sadly an (almost) impossible task for Rust to handle is that there exists a proof that your impl of this trait is actually abelian (rip)

Reply:
Yeah, IDK why I put it this way. Have changed this to an assertion